### ISSUE TITLE
AWS::DynamoDB::Table GlobalSecondaryIndex Support for Multiple GSIs

### ISSUE THREAD

**Issue by thesilentg** [aws-cloudformation/cloudformation-coverage-roadmap#229]
## 2. Scope of request 
Cloudformation cannot perform more than one DynamoDB GSI creation or deletion in a single update. Instead, users must add/delete a single GSI per update, and then manually create a new update for each GSI which needs to be added. For teams which have Cloudformation run as part of a deployment pipeline, this is even more bothersome as it necessitates waiting for multiple end-to-end runs of the entire deployment pipeline for each GSI that needs to be created/deleted. 

Furthermore, once a Cloudformation stack includes a DynamoDB table with more than one GSI, it is no longer possible for someone to re-use this stack without manually editing it and going though the one-at-a-time GSI update process.  This issue seems to be in direct contrast with the advertised goals of CloudFormation. 

https://aws.amazon.com/cloudformation/
> AWS CloudFormation provisions your resources in a safe, repeatable manner, allowing you to build and rebuild your infrastructure and applications, without having to perform manual actions or write custom scripts.

## 3. Expected behavior

1. Update operations to existing DynamoDb tables succeed even when adding/deleting multiple GSIs 
2. Creating a new DynamoDb table using an Cloudformation stack that includes multiple GSIs succeeds 

## 4. Suggest specific test cases

1. Creating a new DynamoDb table with two GSIs is able to complete successfully
2. Adding two GSIs in a single update to an existing table is able to complete successfully
3. Deleting a DynamoDb table with two GSIs is able to complete successfully
4. Removing two GSIs in a single update to an existing table is able to complete successfully

## 5. Helpful Links to speed up research and evaluation

https://stackoverflow.com/questions/28402003/creating-multiple-gsis-by-updatetable-dynamodb
https://stackoverflow.com/questions/47920665/how-to-delete-more-than-one-global-secondary-index-from-the-cloudformation-at-on
[Managing Global Secondary Indexes documentation](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/GSI.OnlineOps.html)
## 6. Category 

DB: DynamoDb

**Comment by PatMyron** [comment]
Creating and deleting multiple DynamoDB tables with multiple Global Secondary Indexes in one operation is possible in CloudFormation already:

[ExampleTemplate.yaml](https://github.com/aws-cloudformation/aws-cloudformation-coverage-roadmap/files/3970867/template.txt)

---

**Difficult to guarantee multiple Global Secondary Index additions in one CloudFormation update for any existing DynamoDB table because:**

1) CloudFormation uses temporary credentials to provision resources
2) DynamoDB Global Secondary Index creation for existing tables is performed serially
3) DynamoDB Global Secondary Index creation time has no guarantees:
> **Adding a Global Secondary Index to a Large Table**
> The time required for building a global secondary index depends on several factors, such as the following:
> * The size of the table
> * The number of items in the table that qualify for inclusion in the index
> * The number of attributes projected into the index
> * The provisioned write capacity of the index
> * Write activity on the main table during index builds
> 
> If you are adding a global secondary index to a very large table, it might take a long time for the creation process to complete.

---

GSI deletion time should be faster, **so possible for CloudFormation to add support for multiple GSI operations on existing tables if one or fewer of the operations are GSI creations**

**Comment by benkehoe** [comment]
Will the new resource framework help solve this, given that it has dedicated support for long-running operations?

**Comment by PatMyron** [comment]
New framework still has time limits for the foreseeable future, right @rjlohan? I don't think there are guarantees on GSI creation time, so that could still be problematic even with longer time limits

**Comment by ikben** [comment]
if I understand it correctly, there is a PR (https://github.com/aws-cloudformation/aws-cloudformation-resource-schema/pull/61) open that would give you a maximum execution time of 12 hours (https://github.com/aws-cloudformation/aws-cloudformation-resource-schema/pull/61/files#diff-c22e96ee3efd905ffecd2b8dcad5c260R23)

**Comment by rjlohan** [comment]
> The new framework will still have some time limit for the foreseeable future, right @rjlohan? I don't think there are any guarantees on GSI creation time, so that case could still be problematic even with a longer time limit

Correct. At some point, we are always going to be bound by the maximum session duration on credentials which CloudFormation assumes from the customer. For private types in the new framework, that maximum is 12 hours, bound by the [maximum duration](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html) we can request in an `sts:AssumeRole` call.

For AWS-owned public types, we use a different authorization mechanism which lets us run to 36hours maximum. Anything longer, or transient, would require a new authZ model inside CloudFormation itself, which is a bigger ask. There are ways to achieve this, but will require some investment.

**Comment by benkehoe** [comment]
I guess when I think about the authentication part of this, which is that the user has an authenticated session that expires at some specific time, I balance that against the purpose of CloudFormation, which is to create persistent resources. It's often going to be the case that I can create resources with CloudFormation that can carry out my session's capability beyond my session expiration anyway.

**Comment by luiseduardocolon** [comment]
We will likely not pursue this in the short- or medium-term.  The API specifically expects to perform only one single creation or deletion per update, and we believe we need to support it as the API dictates it.  I'll close it for now - I believe that if we ever support this it will be if or when the underlying DynamoDB API supports it.

**Comment by et304383** [comment]
This should be reopened.  All AWS needs to do is create a stand-alone resource for GSIs on an existing table.  Then, we could chain them together with a bunch of depends on calls.

**Comment by gelpenaddict** [comment]
I understand that DynamoDB's API allows only 1 GSI but isn't the purpose of CloudFormation to automate deployments, including serializing such API calls to other services?

**Comment by thesilentg** [comment]
Coming back to this almost a year later, after we got hit by the same issue again. 

> The API specifically expects to perform only one single creation or deletion per update, and we believe we need to support it as the API dictates it. 

Cloudformation doesn't expose *any* API, it provides a declarative syntax for defining which resources you want to exist in your AWS account. I'm assuming you're talking about the Dynamo API. When I define a `AWS::DynamoDB::Table` in Cloudformation, the intent is not to call say `Please call CreateTable for me`, it is instead to say "I want a DynamoDb table to exist matching this configuration, do whatever it takes on my behalf to make it so".  To users of Cloudformation, the specifics of what AWS API calls need to get invoked to spin up the resources are an internal implementation detail of Cloudformation which is deliberately abstracted from the user. If I wanted to be invoking `CreateTable`, I wouldn't be writing Cloudformation (I'd be calling `CreateTable` myself). As a declarative tool, Cloudformation makes no statements about how resources come into being, merely what the end state will look like. A DynamoDb table with multiple GSIs is a completely rational end state, therefore I believe it is Cloudformation's duty to translate my intention into existence, regardless of what internal API calls are required to make it so. 

Furthermore, I believe that the validity of a Cloudformation file should be path-independent. That is to say, a given Cloudformation template should be valid (syntactically correct + able to be applied) or invalid (not syntactically correct or not able to be applied) without any ambiguity. To give an example, let's consider five different Cloudformation templates (A - E). 

Cloudformation Templates:
```
A: Empty Template (no resources)
B: Table with no GSIs
C: Table with one GSI
D: Table with two GSIs
E: Table with three GSIs
```

Then lets consider the possible transitions you can take when modifying the Cloudformation template and applying the Cloudformation update: 

Transitions:
```
A -> B: Valid
A -> C: Valid
A -> D: Valid
A -> E: Valid
B -> C: Valid
B -> D: Invalid (trying to add two GSIs)
B -> E: Invalid (trying to add three GSIs)
C -> D: Valid
C -> E:  Invalid (trying to add two GSIs)
D -> E: Valid
```

Now let me pose the following question: Is Cloudformation template D valid or invalid? The answer is of course: "It depends on what state your AWS infrastructure is in already". If you already have a table defined (B), the Cloudformation template D isn't valid. But if you don't have any table defined in your AWS account (A) or you have a table with one GSI already (C), then Cloudformation template D is perfectly valid. 

If I have to know that in order to get from current state B to end state D that I first must take a detour through state C, this is a massive problem.

**Comment by thesilentg** [comment]
> It may be difficult to guarantee multiple Global Secondary Index additions in one CloudFormation update for any existing DynamoDB table because:

>CloudFormation uses temporary credentials to provision resources
>DynamoDB Global Secondary Index creation for existing tables must be performed serially
>DynamoDB Global Secondary Index creation time has no guarantees:

While I understand that creating a GSI is not a bounded operation, and the credentials have a maximum timeout, why can't Cloudformation create _as many GSIs as possible_ within the allowed time limit? Let say the maximum credential session is 12 hours. If I try to create 5 different GSIs on a table in a single cloudformation update and each one takes three hours, I'm fine with this failing (5 * 3 = 15, 15 > 12). But if I try to create three GSIs, why can't all three be updated successfully? What problems does "try to create as many GSIs as possible" cause?

**Comment by ikben** [comment]
> What problems does "try to create as many GSIs as possible
Some problems:
- Rollbacks will be even a bigger pain.
- Applying the same change on two stacks might cause a failure on one (because it is a big table), while it works on the other. Deployments that fail unpredictably are worse than those that always fail.

> If I have to know that in order to get from current state B to end state D that I first must take a detour through state C, this is a problem.
I do agree with this, but with some extra remarks:

- This is currently a per-resource problem, and the possibility for CloudFormation to solve this is dependent on the available service APIs. This twitter thread by @benkehoe is very relevant (see "stage 3"): https://twitter.com/ben11kehoe/status/1278031321550422017 
- Even when the resource would support going directly to every state, you can still cause problems as a template author (eg. If you choose your own TableName, every other deployment of that template, or every replacement of that resource will be impossible)

**Comment by thesilentg** [comment]
> Rollbacks will be even a bigger pain.
Will they? My (admittedly naive) understanding is that rollbacks (GSI deletions) will be much faster than rollouts (GSI creations). Therefore, it shouldn't be possible for a stack to get stuck in a bad state, since even if the GSI creation fails, the rollback will be able to succeed. 

> Applying the same change on two stacks might cause a failure on one (because it is a big table), while it works on the other. Deployments that fail unpredictably are worse than those that always fail.
I'll partially concede this point, although I kind of consider "Deployments that fail unpredictably are worse than those that always fail." to be the situation we have today. When adding additional GSI's to a Cloud formation template, a developer's first deployment will lead to an unpredicted failure until they learn about the single GSI creation restrictions. 

> This twitter thread by @benkehoe is very relevant (see "stage 3"):
This is useful context. 

My end goal here is making it so that developers are able to reliable deploy infrastructure with the minimal amount of manual overhead. What needs to occur in order for this goal to be realized in regards to the ability to update multiple DynamoDB GSIs in a single Cloudformation template change? 

Do I need to push on the Dynamo team to provide an API with support for multiple GSI updates? (that Cloudformation can later consume)

**Comment by gelpenaddict** [comment]
If rollbacks are a pain, then shouldn't the pain be consolidated in CloudFormation rather than distributed to every user facing this scenario?

> A CloudFormation template describes your desired resources and their dependencies so you can launch and configure them together as a stack.

https://aws.amazon.com/cloudformation/

Does the current state with this bug adhere to CloudFormation's description of itself?  Can we describe the desired resources and have it launch as a stack?

**Comment by ikben** [comment]
> [...] rollbacks (GSI deletions) will be much faster than rollouts (GSI creations) [...]

I forgot about the asymmetry there. It will still take a long time before a rollback starts, and It will be non-obvious that the solution is doing the change with one GSI at a time. But it is less of a pain than what I thought. I'd still be wary of possible edge cases (what happens if something else fails together with a delete of an index? ....)

> Do I need to push on the Dynamo team to provide an API with support for multiple GSI updates? (that Cloudformation can later consume)

I don't work for AWS, but from https://github.com/aws-cloudformation/aws-cloudformation-coverage-roadmap/issues/229#issuecomment-573243450, that seems to be the most effective way to get this changed.

**Comment by nubpro** [comment]
> We will likely not pursue this in the short- or medium-term. The API specifically expects to perform only one single creation or deletion per update, and we believe we need to support it as the API dictates it. I'll close it for now - I believe that if we ever support this it will be if or when the underlying DynamoDB API supports it.

Now how do we exactly push DynamoDB team to support this API?

**Comment by gelpenaddict** [comment]
https://aws.amazon.com/cloudformation/
Should probably also update this 
> CloudFormation template describes your desired resources and their dependencies so you can launch and configure them together as a stack

accordingly to something like 

> CloudFormation template is like a script calling other AWS APIs

**Comment by IsaiahJTurner** [comment]
I have proposed an improved solution to this issue over at https://github.com/aws-cloudformation/cloudformation-coverage-roadmap/issues/881
If you are experiencing this issue with DyanamoDB + CloudFormation (as the 50+ upvotes on this post are), take a look at that proposal and give it some love if it would help you too so it can be appropriately prioritized.

**Comment by paul-uz** [comment]
Is there any update on this being supported?

I find it pretty ridiculous that a table can be created with multiple GSI's, but only 1 can be added/removed after the table is created.

**Comment by kuda1992** [comment]
I can't believe this is not fixed yet, this is certainly a deal breaker.

**Comment by clouddev-code** [comment]
In the case of updating multiple GSIs, the Cloudformation stack needs to be run with separate change sets, so I'd like to be able to change them all at once.

**Comment by ahurlburt** [comment]
Just bumping this. Lack of multiple GSI updates really hurts dev time. Current workaround is to update one, deploy, wait (takes 15-20 minutes for our build to build and deploy), repeat.

Then repeat this **per environment again**

It really is cumbersome. Technically it can be worked around but I started using CloudFormation assuming it was going to speed development not hinder it. It wouldn't be as big of a deal if the deploys didn't take so long.

**Comment by paul-uz** [comment]
I still fail to see the technical reasoning for why this can't be done? Surely there isn't one?

I'd happily forfeit a bit of additional deployment time if it meant it worked.

**Comment by bfbenf** [comment]
I am also facing the same issue and like @paul-uz fail to see why this cannot be handled by CloudFormation. I am using CDK to produce our CloudFormation and commenting GSI's make our tests fail in the pipeline and makes deploying Tables really cumbersome

**Comment by PeterBaker0** [comment]
Adding another voice to the consensus above - this completely violates the normal CDK -> CloudFormation -> Pipeline workflow. I think the illustrative and detailed example above regarding the definition of "valid" and "invalid" is spot on - unless absolutely unavoidable, the validity of a template should not depend on the current state of the infrastructure.

**Comment by wz2b** [comment]
> We will likely not pursue this in the short- or medium-term. The API specifically expects to perform only one single creation or deletion per update, and we believe we need to support it as the API dictates it.

You're right that the [UpdateTable](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_UpdateTable.html) supports only one deletion or modification per update, but [CreateTable](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_CreateTable.html) supports up to 20.  So I think something might be awry in CloudFormation.

**Comment by christophermichaelthomasmillar** [comment]
The lack of support for something so fundamental really speaks against using DynamoDB with cloud formation in general. I feel like in all the cloud hype we forgot the difference between good software and bad and now we're left SOL. Like a branch Davidian with a stomach ache I'm beginning to question my life choices in joining this cargo cult.

**Comment by mobob** [comment]
AWS: _hey so we got this great gsi feature, it lets you support all your data access patterns in dynamo making it an actually usable database_

Everyone: _cool, can't wait to kick it!!_

Everyone: _ok now i've got two additional access patterns and i'm trying to test them but I'm hitting this issue...  is this really a limitation, or am i doing something wrong?_

AWS: _sorry, follow some link to a npm package that is dead but that is a preferred solution, or some other convoluted approach that may might work someday_

Everyone: _whoa that sounds weird. maybe its only happening cause i'm doing dev, some people suggest its only during mutation of the indexes, surely once i push to prod and have 2 or more GSIs with no changes to them that'll work_

Everyone: *goes to prod* - _nope, still fails...  wtf, how is this possible?_

Hmmm ok i'm probably missing something, but this is how I feel, today, as i pushed 5 GSI's to prod. Sad. Am i really going to incrementally deply them by...  checking one in at a time? I _must_ be missing something.

**Comment by paul-uz** [comment]
@mobob you can create a table with 5 GSIs, you just can't add 4 GSIs to a table that already has one, in one go. Utter nonsense

**Comment by mobob** [comment]
> @mobob you can create a table with 5 GSIs, you just can't add 4 GSIs to a table that already has one, in one go. Utter nonsense

Right - thanks for the clarification.

+1 to getting some solution that allows incremental GSI changes to dynamo db.

**Comment by ahurlburt** [comment]
yea this is a huge pain to do a bunch of deployments to add 1 at a time, especially since deployments take so long.

This thread has a huge amount of likes, maybe eventually a PM will take notice.

**Comment by Ccz-Chen** [comment]
Any update/progress on this issue? We have to run N# deployments in order to push a feature to public, if the feature uses N# of GISs because of this error message:

"Resource handler returned message: "Cannot perform more than one GSI creation or deletion in a single update" (RequestToken: 4c2ffec4-a684-dac1-0478-6cc98d09c044, HandlerErrorCode: InvalidRequest)"

**Comment by BryanCrotaz** [comment]
The best way for CF to fix this is to add a new resource type for Dynamo GSI that can be attached to a { Ref: DynamoTable }
Then CF can know to run them sequentially, and if that will take too long for the credentials, start a step function behind the scenes that runs each one, then the rest of the script.

**Comment by soplan** [comment]
Yes feeling the same pain here. We have 3 different stages. Usually on Dev we deploy multiple times a day so we are not hindered by this issue. But once we are done testing and ready to merge to staging we are stuck because of this issue. We want to do CI/CD without headaches.

**Comment by leonardogazdek** [comment]
Ditched CloudFormation entirely because of this. For me this pretty much defeats the whole purpose of CloudFormation with DynamoDB.

**Comment by paul-uz** [comment]
> Ditched CloudFormation entirely because of this. For me this pretty much defeats the whole purpose of CloudFormation with DynamoDB. 

What did you replace CFN with?

**Comment by pauldraper** [comment]
> I'll close it for now - I believe that if we ever support this it will be if or when the underlying DynamoDB API supports it.

@luiseduardocolon you might consider talking to the Terraform team.

They were clever enough to figure out a solution.

Couldn't hurt to get some help. 🤷

**Comment by CharlieGreenman** [comment]
Hey I now it's closed, but I was wondering if any further discussion in this regard? For myself I use a versioning approach. Maybe AWS can version indexes i.e. create a new one, and then cutover. This is how our team deals with at the moment: https://charliegreenman.medium.com/using-versioning-for-aws-gsis-20ecadd3c53c

**Comment by PatMyron** [comment]
> > Creating and deleting multiple DynamoDB tables with multiple Global Secondary Indexes in one operation is possible in CloudFormation already:
> > [ExampleTemplate.yaml](https://github.com/aws-cloudformation/aws-cloudformation-coverage-roadmap/files/3970867/template.txt)
> 
> I can't believe someone from the CFN team suggested this as an example. The issue is clearly about defining 2 or more GSIs on a single table and you give a template with two table resources each with one GSI.

@et304383 both tables have multiple GSIs.. read carefully before mocking the 1st person advocating for an issue ;)

**Comment by PeterBaker0** [comment]
> Hey I now it's closed, but I was wondering if any further discussion in this regard? For myself I use a versioning approach. Maybe AWS can version indexes i.e. create a new one, and then cutover. This is how our team deals with at the moment: https://charliegreenman.medium.com/using-versioning-for-aws-gsis-20ecadd3c53c

Hi @CharlieGreenman - does this actually address the problem? What if you want to add two new versions of a GSI in the same deployment? e.g. emailIndex-v2 and usernameIndex-v2. Wouldn't this require a) delete two existing indexes b) produce two new indexes (depending on if you replace or append the indexes).

**Comment by CharlieGreenman** [comment]
> > Hey I now it's closed, but I was wondering if any further discussion in this regard? For myself I use a versioning approach. Maybe AWS can version indexes i.e. create a new one, and then cutover. This is how our team deals with at the moment: https://charliegreenman.medium.com/using-versioning-for-aws-gsis-20ecadd3c53c
> 
> Hi @CharlieGreenman - does this actually address the problem? What if you want to add two new versions of a GSI in the same deployment? e.g. emailIndex-v2 and usernameIndex-v2. Wouldn't this require a) delete two existing indexes b) produce two new indexes (depending on if you replace or append the indexes). 


It works. Create new index in template yaml. Wait until backfill complete. Then update variable for index in cloudformation so new index. Once use new GSI can remove the old one. 

It's a way of making sure prod GSIs always stay up

**Comment by PeterBaker0** [comment]
> Create new index in template yaml.

What if you create two?

**Comment by CharlieGreenman** [comment]
> > Create new index in template yaml.
> 
> What if you create two?

It works because creating them from scratch. From the aws side, they need a webhook to determine when backfilling for new version complete, and then cut over. 

If the AWS team does this, they can solve the GSI issue. They just need to version GSIs

**Comment by PeterBaker0** [comment]
Sure - the AWS CloudFormation team could easily solve this problem - lots of options! Sorry I just meant that for readers of this thread - your approach doesn't actually mitigate the problem raised in this issue- it still requires many redeployments of each GSI sequentially (no matter your naming scheme). 

Appreciate this thread being raised regardless - insanity this isn't resolved yet...

**Comment by CharlieGreenman** [comment]
> Sure - the AWS CloudFormation team could easily solve this problem - lots of options! Sorry I just meant that for readers of this thread - your approach doesn't actually mitigate the problem raised in this issue- it still requires many redeployments of each GSI sequentially (no matter your naming scheme). 
> 
> Appreciate this thread being raised regardless - insanity this isn't resolved yet...

It solves the "keeping prod at 100% without downtime" issue. If delete and re-add index will backfill for a while.
